packages/from-hardhat/src/compilation.ts

```diff
@@ -0,0 +1,121 @@
+import type * as Hardhat from "hardhat/types";
+
+import type { Compilation, CompiledContract } from "@truffle/compile-common";
+import * as CompileSolidity from "@truffle/compile-solidity";
+
+import { supportedHardhatBuildInfoFormats } from "./constants";
+
+export const fromBuildInfo = (buildInfo: Hardhat.BuildInfo): Compilation => {
+  const { _format } = buildInfo;
+
+  if (!supportedHardhatBuildInfoFormats.has(_format)) {
+    throw new Error(`Unsupported build info format: ${_format}`);
+  }
+
+  const sourceIndexes = SourceIndexes.fromBuildInfo(buildInfo);
+
+  return {
+    sourceIndexes,
+    sources: Sources.fromBuildInfo(buildInfo, sourceIndexes),
+    contracts: Contracts.fromBuildInfo(buildInfo),
+    compiler: {
+      name: "solc",
+      version: buildInfo.solcLongVersion
+    }
+  };
+};
+
+namespace SourceIndexes {
```

Comment:
Hmm, so I'm not sure how I feel about the use of namespaces in TS code. I generally regard them as strictly for the purpose of writing type declarations for pure JS modules that do complicated things that would otherwise be inexpressible via the TS type system.

My main concern with them is pretty minor, however. Basically it just feels a bit odd reading `SourceIndexes.fromBuildInfo` in the code above, then discovering that it's not a static method on a class, but rather a namespaced local function.

I think also given that the namespaces here aren't exported, it _really_ is quite a minor matter of stylistic opinion.

Reply:
Yeah I did this only for stylistic reasons (I like the function name consistency `fromBuildInfo`, and I like that it means I can use `export const fromBuildInfo = () => new Promise(/* ... */)`)